compiler/rustc_parse/src/parser/path.rs

```diff
@@ -624,9 +624,18 @@ impl<'a> Parser<'a> {
             GenericArg::Const(self.parse_const_arg()?)
         } else if self.check_type() {
             // Parse type argument.
+            let is_const_fn = self.look_ahead(1, |t| t.kind == token::OpenDelim(token::Paren));
+            let mut snapshot = self.clone();
             match self.parse_ty() {
                 Ok(ty) => GenericArg::Type(ty),
                 Err(err) => {
+                    if is_const_fn {
+                        if let Ok(expr) = snapshot.parse_expr_res(Restrictions::CONST_EXPR, None) {
```

Comment:
It looks like from the error that you're getting that attempting this additional parsing might be producing duplicate errors for some cases?

Reply:
After debugging, I found that cloning the parser causes the problem. If you try the following, the error is duplicated.

```rust
} else if self.check_type() {
    // Parse type argument.
    // let is_const_fn = self.look_ahead(1, |t| t.kind == token::OpenDelim(token::Paren));
    let _snapshot = self.clone();
    match self.parse_ty() {
        Ok(ty) => GenericArg::Type(ty),
        Err(err) => {
            // if is_const_fn {
            //     if let Ok(expr) = snapshot.parse_expr_res(Restrictions::CONST_EXPR, None) {
            //         *self = snapshot;
            //         return Ok(Some(self.dummy_const_arg_needs_braces(err, expr.span)));
            //     }
            // }

            // Try to recover from possible `const` arg without braces.
            return self.recover_const_arg(start, err).map(Some);
        }
    }
} else if self.token.is_keyword(kw::Const) {
```

Reply:
After debugging, I found that cloning the parser causes the problem. If you try the following, the error is duplicated.

```rust
} else if self.check_type() {
    // Parse type argument.
    // let is_const_fn = self.look_ahead(1, |t| t.kind == token::OpenDelim(token::Paren));
    let _snapshot = self.clone();
    match self.parse_ty() {
        Ok(ty) => GenericArg::Type(ty),
        Err(err) => {
            // if is_const_fn {
            //     if let Ok(expr) = snapshot.parse_expr_res(Restrictions::CONST_EXPR, None) {
            //         *self = snapshot;
            //         return Ok(Some(self.dummy_const_arg_needs_braces(err, expr.span)));
            //     }
            // }

            // Try to recover from possible `const` arg without braces.
            return self.recover_const_arg(start, err).map(Some);
        }
    }
} else if self.token.is_keyword(kw::Const) {
```

Reply:
Hmm. I've debugged a lot of things and can't understand why cloning the parser causes duplicated errors.

Reply:
Cloning the parser before calling `parse_xxx` other than `parse_ty` does't cause this problem.

Reply:
I would appreciate any solutions or other suggestions.

Reply:
Looks like `Parser` has a `Drop` implementation that emits errors based on the contents of the [`unclosed_delims`](https://doc.rust-lang.org/stable/nightly-rustc/rustc_parse/parser/struct.Parser.html#structfield.unclosed_delims) field and your snapshot is emitting those when dropped? That seems easy enough to confirm.

If that is the case, perhaps make a `diagnostic_snapshot` function on `Parser` which does a `clone` and then clears any state like this (probably just `unclosed_delims`) and use that anywhere that snapshots like these are being made. In fact, perhaps that isn't a great idea because if we want the snapshot parser to become the actual parser (i.e. `*self = snapshot`) then we don't want to lose anything we might want if the snapshot is to be kept - maybe you can experiment with adding a nice API for this that lets us stash away that state that could produce errors when the snapshot gets dropped without throwing away that state entirely, just in case we want to keep the snapshot.

Reply:
Thank you. I missed the `Drop` implementation.